timelock/program/src/utils.rs

```diff
@@ -0,0 +1,414 @@
+use crate::{
+    error::TimelockError,
+    state::{
+        enums::TimelockStateStatus, timelock_program::TimelockProgram, timelock_set::TimelockSet,
+        timelock_state::TimelockState,
+    },
+};
+use arrayref::array_ref;
+use solana_program::{
+    account_info::AccountInfo,
+    entrypoint::ProgramResult,
+    instruction::Instruction,
+    program::invoke_signed,
+    program_error::ProgramError,
+    program_pack::{IsInitialized, Pack},
+    pubkey::Pubkey,
+    system_instruction::create_account,
+    sysvar::rent::Rent,
+};
+use spl_token::state::Account;
+
+/* TODO come back to this conundrum later..
+
+pub fn get_authority_signer_seeds<'a>(
+    timelock_program_account_info: &'a AccountInfo<'a>,
+    timelock_program_authority_info: &'a AccountInfo<'a>,
+    program_id: &'a Pubkey,
+) -> Result<&'a [&'a [u8]; 2], ProgramError> {
+    let (authority_key, bump_seed) =
+        Pubkey::find_program_address(&[timelock_program_account_info.key.as_ref()], program_id);
+    if timelock_program_authority_info.key != &authority_key {
+        return Err(TimelockError::InvalidTimelockAuthority.into());
+    }
+    let authority_signer_seeds = &[timelock_program_account_info.key.as_ref(), &[bump_seed]];
+    Ok(&*authority_signer_seeds)
+}*/
+
+/// Attempts to transfer the token to the timelock set's validation account and back to the person again.
+/// Can only be done if done in a transaction that has authority to do so. Serves as a check
+/// That the person is who they say they are!
+pub fn assert_is_permissioned<'a>(
+    program_id: &Pubkey,
+    perm_account_info: &AccountInfo<'a>,
+    perm_validation_account_info: &AccountInfo<'a>,
+    timelock_program_info: &AccountInfo<'a>,
+    token_program_info: &AccountInfo<'a>,
+    transfer_authority_info: &AccountInfo<'a>,
+    timelock_authority_info: &AccountInfo<'a>,
+) -> ProgramResult {
+    let _perm_account: Account = assert_initialized(perm_account_info)?;
+    let _perm_validation: Account = assert_initialized(perm_validation_account_info)?;
+    let (authority_key, bump_seed) =
+        Pubkey::find_program_address(&[timelock_program_info.key.as_ref()], program_id);
+    if timelock_authority_info.key != &authority_key {
+        return Err(TimelockError::InvalidTimelockAuthority.into());
+    }
+    let authority_signer_seeds = &[timelock_program_info.key.as_ref(), &[bump_seed]];
+    // If both accounts arent correct mint type, it explodes
```

Comment:
Cool idea!  This has some unclear behavior for delegations that we need to watch.  If I delegate my admin token to someone else, and then they call this on my behalf, that delegation will disappear without me having revoked it.

Reply:
Well, that's the point isn't it, right? We want the admin token to be the sole proprietorship. If they have your admin token, they can do whatever they want. I'm not sure what the problem is?

Reply:
I'm mentioning something a bit different, which might end up being fine.  Along with checking all of the mints for you, this also has the neat feature of working with `approve`s.  So if I delegate my admin token to ephemeral K and sign with it, it would work for this.  On the flip side, since the admin token is transferred here using K, the delegation to K would expire after this instruction.

Reply:
When you delegate your token to K, is this different than transferring it? Is this a special transaction-scoped command for spl-token I don't yet know about?

Reply:
Yes this is different, when using the term `delegate` I've meant the `approve` instruction on spl-token: https://github.com/solana-labs/solana-program-library/blob/master/token/program/src/instruction.rs#L109

This allows people to effectively scope how much of their tokens can get taken by a program.  I `approve` a small amount of my tokens to an ephemeral key pair, then sign with the ephemeral keypair.  Even if the program is malicious, the most it could take is that small amount of tokens.  On the other hand, without an `approve` instruction, I could sign a transaction with my account, and the program could take all of my tokens.

Reply:
Oh, I understand. K could then have permission to transfer the admin token back and forth. I got it.